src/puppetlabs/puppetdb/command.clj

```diff
@@ -213,13 +214,18 @@
 
 ;; Node deactivation
 
-(defmethod process-command! [(command-names :deactivate-node) 2]
-  [{certname :payload {:keys [id]} :annotations} {:keys [db]}]
-  (jdbc/with-transacted-connection db
-    (when-not (scf-storage/certname-exists? certname)
-      (scf-storage/add-certname! certname))
-    (scf-storage/deactivate-node! certname))
-  (log/info (format "[%s] [%s] %s" id (command-names :deactivate-node) certname)))
+(defmethod process-command! [(command-names :deactivate-node) 3]
+  [{:keys [payload annotations]} {:keys [db]}]
+  (let [certname (:certname payload)
+        producer_timestamp (to-timestamp (:producer_timestamp payload (now)))
+        id  (:id annotations)
+        have-newer? (fn [entity] (scf-storage/have-record-produced-after? entity certname producer_timestamp))]
+    (jdbc/with-transacted-connection db
+      (when-not (scf-storage/certname-exists? certname)
+        (scf-storage/add-certname! certname))
+      (when-not (or (have-newer? :catalogs) (have-newer? :factsets) (have-newer? :reports))
```

Comment:
What would you think about `(some have-newer? [:catalogs :factsets :reports])` here?